### PR TITLE
Anti Mage Lua

### PR DESCRIPTION
**Pull request by JoyfulShush**

### CODE REVIEW

File: game/scripts/vscripts/heroes/antimage/antimage_mana_break.lua [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    self.hit_sound = "Hero_Antimage.ManaBreak"
+    self.particle_manaburn = "particles/generic_gameplay/generic_manaburn.vpcf"
+
+    -- Ability specials
+    self.damage_per_burn = self.ability:GetSpecialValueFor("damage_per_burn")
+    self.mana_per_hit = self.ability:GetSpecialValueFor("mana_per_hit")
+end
+
+function modifier_lua_antimage_mana_break:OnRefresh()
+    self:OnCreated()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_mana_break.lua]: eeeeeeeeh, better to just update the special values in here than rerun the oncreated.

File: game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+LinkLuaModifier("modifier_lua_mana_void_stunned", "antimage/antimage_mana_void", LUA_MODIFIER_MOTION_NONE)
+
+function lua_antimage_mana_void:GetAbilityTextureName()
+    return "antimage_mana_void"
+end
+
+function lua_antimage_mana_void:OnAbilityPhaseStart()
+    if IsServer() then
+        self:GetCaster():EmitSound("Hero_Antimage.ManaVoidCast")
+        return true
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua]: Should this return not be outside of the server block?

File: game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        ParticleManager:SetParticleControlEnt(void_pfx, 0, target, PATTACH_POINT_FOLLOW, "attach_hitloc", target:GetOrigin(), true)
+        ParticleManager:SetParticleControl(void_pfx, 1, Vector(mana_void_aoe_radius,0,0))
+        ParticleManager:ReleaseParticleIndex(void_pfx)
+        target:EmitSound("Hero_Antimage.ManaVoid")    
+    end 
+end
+
+
+-- Stun modifier
+modifier_lua_mana_void_stunned = class({})
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua:96]: why not just use modifier_stunned?

File: game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+    -- Ability specials
+    self.spell_shield_resistance = self.ability:GetSpecialValueFor("spell_shield_resistance")    
+
+    -- Initialize table of old spells
+    self.caster.tOldSpells = {}
+    
+    if IsServer() then
+        -- Think, in order to check for scepter. Once a scepter is found, the modifier's interval starts
+        Timers:CreateTimer(function()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua]: we should probably try to limit the usage of Timers, in modifiers you can simply do `self:StartIntervalThink()` and `modifier_lua_antimage_spell_shield:OnIntervalThink()`

File: game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                self:StartIntervalThink(0.2)
+            else
+                return 20
+            end
+        end)
+    end
+end
+
+function modifier_lua_antimage_spell_shield:OnRefresh()
+    self:OnCreated()
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua]: I wouldn't run oncreated again, that is for **creation**, just update whatever you need in here.

File: game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        if self.caster:HasScepter() and self.caster:IsRealHero() and self.ability:IsCooldownReady() then
+            if not self.caster:PassivesDisabled() then  
+                -- Apply Spell Absorption
+                local reflect_pfx = ParticleManager:CreateParticle("particles/units/heroes/hero_antimage/antimage_spellshield.vpcf", PATTACH_CUSTOMORIGIN_FOLLOW, self.caster)
+                ParticleManager:SetParticleControlEnt(reflect_pfx, 0, self.caster, PATTACH_POINT_FOLLOW, "attach_hitloc", self.caster:GetOrigin(), true)
+                ParticleManager:ReleaseParticleIndex(reflect_pfx)
+                return 1
+            end
+        end
+        return false
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua]: Either return true and false or 1 and 0, not 1 and false

File: game/vscripts/npc/npc_abilities_custom.txt [hunk 7]
```diff
@@ -0,0 +1,110 @@
+"DOTAAbilities"
```
Comment (Perryvw) [inline game/vscripts/npc/npc_abilities_custom.txt]: Don't think this file is supposed to be in this pull request.

File: game/vscripts/scripts/antimage/antimage_blink.lua [hunk 8]
```diff
@@ -0,0 +1,64 @@
+-- Author: Shush
+-- Date: 25/11/2017
+
+-------------------------------
+--   Anti Mage's Blink  --
+-------------------------------
+
```
Comment (Perryvw) [inline game/vscripts/scripts/antimage/antimage_blink.lua]: Duplicate file

File: game/vscripts/scripts/antimage/antimage_mana_break.lua [hunk 9]
```diff
@@ -0,0 +1,103 @@
+-- Author: Shush
+-- Date: 25/11/2017
+
+-------------------------------
+--   Anti Mage's Mana Break  --
+-------------------------------
+
```
Comment (Perryvw) [inline game/vscripts/scripts/antimage/antimage_mana_break.lua]: Duplicate file

File: game/vscripts/scripts/antimage/antimage_spell_shield.lua [hunk 10]
```diff
@@ -0,0 +1,43 @@
+-- Author: Shush
+-- Date: 25/11/2017
+
+-------------------------------
+--  Anti Mage's Spell Shield --
+-------------------------------
+
```
Comment (Perryvw) [inline game/vscripts/scripts/antimage/antimage_spell_shield.lua]: Duplicate file.

**Review by Perryvw — CHANGES_REQUESTED** [review]
Looks good overall, requested some small changes and removal of duplicate file.

File: game/scripts/vscripts/heroes/antimage/antimage_mana_break.lua (hunk 1 above)
Comment (JoyfulShush) [reply game/scripts/vscripts/heroes/antimage/antimage_mana_break.lua]: Agreed.

File: game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua (hunk 2 above)
Comment (JoyfulShush) [reply game/scripts/vscripts/heroes/antimage/antimage_mana_void.lua]: The IsServer() isn't supposed to be there at all as OnAbilityPhaseStart is only called server-side anyway.

File: game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua (hunk 4 above)
Comment (JoyfulShush) [reply game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua]: Agreed. Will rework it soon.

File: game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua (hunk 6 above)
Comment (JoyfulShush) [reply game/scripts/vscripts/heroes/antimage/antimage_spell_shield.lua]: Agreed, copy pasted from imba's stuff.

**Review by SinZ163 — APPROVED** [review]

**Pull request merged**